pkg/entities/record.go

```diff
@@ -102,16 +99,17 @@ func (b *baseRecord) GetFieldCount() uint16 {
 	return b.fieldCount
 }
 
-func (d *baseRecord) GetOrderedElementList() []*InfoElementWithValue {
-	return d.orderedElementList
+func (b *baseRecord) GetOrderedElementList() []*InfoElementWithValue {
+	return b.orderedElementList
 }
 
 func (b *baseRecord) GetInfoElementWithValue(name string) (*InfoElementWithValue, bool) {
-	if element, exist := b.elementsMap[name]; exist {
-		return element, exist
-	} else {
-		return nil, false
+	for _, element := range b.orderedElementList {
+		if element.Element.Name == name {
+			return element, true
+		}
 	}
```

Comment:
Adding the recent profiling result and it shows in top10 of time consumed when I tried this patch. Sharing it here for reference:
```
(pprof) top 20
Showing nodes accounting for 42.71s, 71.20% of 59.99s total
Dropped 343 nodes (cum <= 0.30s)
Showing top 20 nodes out of 122
      flat  flat%   sum%        cum   cum%
    16.09s 26.82% 26.82%     16.22s 27.04%  syscall.Syscall
     5.36s  8.93% 35.76%     11.11s 18.52%  runtime.scanobject
     3.42s  5.70% 41.46%     14.25s 23.75%  runtime.mallocgc
     2.60s  4.33% 45.79%      3.37s  5.62%  runtime.findObject
     1.89s  3.15% 48.94%     10.20s 17.00%  github.com/vmware/go-ipfix/pkg/entities.(*dataRecord).AddInfoElement
     1.74s  2.90% 51.84%      2.29s  3.82%  github.com/vmware/go-ipfix/pkg/entities.(*baseRecord).GetInfoElementWithValue
```
```
      20ms       20ms    103:func (b *baseRecord) GetInfoElementWithValue(name string) (*InfoElementWithValue, bool) {
     610ms      620ms    104:	for _, element := range b.orderedElementList {
     1.16s      1.49s    105:		if element.Element.Name == name {
      40ms       40ms    106:			return element, true
         .          .    107:		}
         .          .    108:	}
         .          .    109:	return nil, false
         .          .    110:}
```

Reply:
I can see the issue in maintaining a sorted list here.
Could you remind me why we need an ordered list? Is it follow the order of template elements when encoding the data record?

Reply:
I was about to mention that. If we want to do binary search, we need to ensure the elements users added are in some order (e.g. alphabetically) otherwise sorting will add time complexity. Yes, we need to keep the ordered list because elements in data record should be in same order as template.

Reply:
I see that Antrea flow exporter maintains its own element list order. However, flow aggregator uses `GetOrderedElementList`.
I understand ordered map is one option. What do you think of following data structure with map?
Have a map where the key is element name and value is a struct with `InfoElementWithValue` and `fieldCount(index)`.

As part of `GetOrderedElementList`, we can create a list by iterating over the map according to each element index.
Could we benchmark this? I can also run the flow aggregator perf test with this patch.
I am concerned about the compute time overhead with the above ` GetInfoElementWithValue` overhead when using the flow aggregator process.

Reply:
Inspired from your suggestion, I tried with a little different way where we keep a map of name-index and an ordered element list so that we can return `GetOrderedElementList` and `GetInfoElementWithValue` within O(1) time complexity. There is not much difference at my side. I think it is because flow exporter does not use these two functions much. Please try benchmarking on your side to see whether it will improve. Thanks!

Reply:
Thanks for trying.
I tried both patches with the fix and without the fix. Moving to map actually degraded the performance (compute time) because of overhead in map access etc. And memory consumption has increased too. Let's keep the list and O(n) implementation of `GetInfoElementWithValue`
Without fix:
```
vagrant@k8s-node-worker-1:~/antrea/pkg/flowaggregator$ go test -test.v -run=none -test.benchmem  -bench=. -memprofile memprofile.out -cpuprofile profile.out
goos: linux
goarch: amd64
pkg: antrea.io/antrea/pkg/flowaggregator
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkIntraNodeFlowRecords
    flowaggregator_perf_test.go:446: Num messages received: 1550292
BenchmarkIntraNodeFlowRecords-2   	       1	120017706287 ns/op	10968399688 B/op	324291038 allocs/op
PASS
ok  	antrea.io/antrea/pkg/flowaggregator	120.204s
```
With fix:
```
vagrant@k8s-node-worker-1:~/antrea/pkg/flowaggregator$ go test -test.v -run=none -test.benchmem  -bench=. -memprofile memprofile.out -cpuprofile profile.out
goos: linux
goarch: amd64
pkg: antrea.io/antrea/pkg/flowaggregator
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkIntraNodeFlowRecords
    flowaggregator_perf_test.go:446: Num messages received: 1508983
BenchmarkIntraNodeFlowRecords-2   	       1	120005281632 ns/op	12688787552 B/op	326282797 allocs/op
PASS
ok  	antrea.io/antrea/pkg/flowaggregator	120.214s
```

Reply:
Got it. Thanks!